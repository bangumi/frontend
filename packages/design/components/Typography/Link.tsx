import classNames from 'classnames';
import type { PropsWithChildren } from 'react';
import React from 'react';
import type { LinkProps as RouterLinkProps } from 'react-router-dom';
import { Link as RouterLink } from 'react-router-dom';

interface PureLinkProps {
  isExternal?: boolean;
  fontWeight?: 'bold';
  noStyle?: boolean;
}

export type LinkProps = PureLinkProps & RouterLinkProps;

/**
 * 有些场景，to 属性无法覆盖到，比如返回上一页，我们需要 navigate(-1)，可我们又需要 Link 的样式
 * 纯 Link 组件就是仅保留了样式，而无任何逻辑。
 */
export function PureLink({
  children,
  className,
  fontWeight,
  ...rest
}: PropsWithChildren<PureLinkProps & JSX.IntrinsicElements['div']>) {
  return (
    <div
      className={classNames(
        'bgm-link',
        {
          'bgm-link--bold': fontWeight === 'bold',
        },
        className,
      )}
      {...rest}
    >
      {children}
    </div>
  );
}

const Link: React.FC<LinkProps> = ({
  to,
  className,
  children,
  fontWeight,
  isExternal = false,
  noStyle = false,
  ...rest
}) => {
  const resolvedClassnames = noStyle
    ? className
    : classNames(
        'bgm-link',
        {
          'bgm-link--bold': fontWeight === 'bold',
        },
        className,
      );

  if (isExternal && typeof to === 'string') {
    return (
      <a className={resolvedClassnames} href={to} {...rest}>
        {children}
      </a>
    );
  }

  return (
    <RouterLink className={resolvedClassnames} to={to} {...rest}>
      {children}
    </RouterLink>
  );
};

export default Link;
