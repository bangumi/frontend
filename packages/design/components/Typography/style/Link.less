@import (reference) '../../../theme/base';

.bgm-link {
  display: inline-block;
  color: @blue-100;
  cursor: pointer;

  &:link,
  &:visited,
  &:active {
    text-decoration: none;
  }

  &:hover {
    text-decoration: underline;
  }

  &--bold {
    font-weight: bold;
  }
}
