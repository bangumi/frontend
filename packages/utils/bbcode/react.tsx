import React from 'react';

import { convert } from './convert';
import { Parser } from './parser';
import type { CodeNodeTypes, NodeTypes, VNode } from './types';

/* eslint-disable no-useless-escape */
const toCamelCase = (s: string) => s.replace(/(\-\w)/g, (k) => k[1]?.toUpperCase() ?? '');

const joinClassName = (cls: string | string[]) => {
  if (Array.isArray(cls)) {
    return cls.join(' ');
  }
  return cls;
};

const convertToReactProps = (node: VNode, key?: React.Key) => {
  const { props, style, className } = node;

  const transformedStyle: Record<string, string> = style
    ? Object.entries(style).reduce((pre, [key, val]) => {
        return {
          ...pre,
          [toCamelCase(key)]: val,
        };
      }, {})
    : {};

  const transformedClassName = className !== undefined ? joinClassName(className) : undefined;

  const reactProps = {
    ...props,
    style: transformedStyle,
    className: transformedClassName,
    key,
  };

  return reactProps;
};

export const renderNode = (node: NodeTypes, key?: React.Key): React.ReactElement | string => {
  if (typeof node === 'string') {
    const splittedStr: Array<string | React.ReactElement> = node.split(/\n/g);

    if (splittedStr.length < 2) {
      return splittedStr[0] ?? '';
    }
    const children = splittedStr.reduce<Array<string | React.ReactElement>>((pre, curr, idx) => {
      if (idx > 0) {
        return [...pre, React.createElement('br', null), curr];
      }
      return [...pre, curr];
    }, []);
    return React.createElement(React.Fragment, { key }, ...children);
  }

  const { type, children } = node;

  if (children) {
    return React.createElement(
      type,
      convertToReactProps(node, key),
      ...children.map((child, idx) => renderNode(child, idx)),
    );
  }
  return React.createElement(type, convertToReactProps(node, key));
};

export const render = (rawStr: string): Array<React.ReactElement | string> => {
  const nodes: CodeNodeTypes[] = new Parser(rawStr).parse();

  return nodes.map((node, idx) => renderNode(convert(node, {}), idx));
};
