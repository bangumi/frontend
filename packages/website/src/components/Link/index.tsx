import type { PropsWithChildren } from 'react';
import React, { createContext, forwardRef, useContext, useTransition } from 'react';
import type { LinkProps, NavLinkProps } from 'react-router-dom';
import { useHref, useLinkClickHandler, useLocation, useResolvedPath } from 'react-router-dom';

interface ILinkContext {
  pending: boolean;
  startTransition: React.TransitionStartFunction;
}
const LinkContext = createContext<ILinkContext>(null!);

export const LinkProvider = ({ children }: PropsWithChildren) => {
  const [pending, startTransition] = useTransition();
  return (
    <LinkContext.Provider
      value={{
        pending,
        startTransition,
      }}
    >
      {children}
    </LinkContext.Provider>
  );
};

export const Link = forwardRef<HTMLAnchorElement, LinkProps>(function LinkWithRef(
  { to, replace, state, target, preventScrollReset, relative, onClick, ...rest },
  ref,
) {
  const { startTransition } = useContext(LinkContext);
  const href = useHref(to, { relative });
  const internalOnClick = useLinkClickHandler(to, {
    replace,
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    state,
    target,
    preventScrollReset,
    relative,
  });
  const handleOnClick: React.MouseEventHandler<HTMLAnchorElement> = (event) => {
    onClick?.(event);
    startTransition(() => {
      internalOnClick(event);
    });
  };
  return <a {...rest} href={href} onClick={handleOnClick} ref={ref} />;
});

export const useLinkContext = () => useContext(LinkContext);

/**
 * NavLink
 * https://github.com/remix-run/react-router/blob/main/packages/react-router-dom/index.tsx#L470
 */
export const NavLink = React.forwardRef<HTMLAnchorElement, NavLinkProps>(function NavLinkWithRef(
  {
    'aria-current': ariaCurrentProp = 'page',
    caseSensitive = false,
    className: classNameProp = '',
    end = false,
    style: styleProp,
    to,
    children,
    ...rest
  },
  ref,
) {
  const location = useLocation();
  const path = useResolvedPath(to);
  const { pending } = useLinkContext();

  let locationPathname = location.pathname;
  let toPathname = path.pathname;
  if (!caseSensitive) {
    locationPathname = locationPathname.toLowerCase();
    toPathname = toPathname.toLowerCase();
  }

  const isActive =
    locationPathname === toPathname ||
    (!end &&
      locationPathname.startsWith(toPathname) &&
      locationPathname.charAt(toPathname.length) === '/');

  const ariaCurrent = isActive ? ariaCurrentProp : undefined;

  return (
    <Link {...rest} aria-current={ariaCurrent} ref={ref} to={to}>
      {typeof children === 'function' ? children({ isActive, isPending: pending }) : children}
    </Link>
  );
});
