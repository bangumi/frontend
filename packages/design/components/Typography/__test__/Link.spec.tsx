import { render } from '@testing-library/react';
import type { PropsWithChildren } from 'react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';

import Link from '../Link';

const LinkTestWrapper: React.FC<PropsWithChildren<{}>> = ({ children }) => {
  return <MemoryRouter>{children}</MemoryRouter>;
};

it('should render internal link', () => {
  const { asFragment } = render(<Link to='/a' />, { wrapper: LinkTestWrapper });

  expect(asFragment()).toMatchSnapshot();
});

it('should render external link', () => {
  const { asFragment } = render(<Link to='https://test.com' isExternal />, {
    wrapper: LinkTestWrapper,
  });

  expect(asFragment()).toMatchSnapshot();
});

it('should render no-style link', () => {
  const { asFragment } = render(<Link to='/a' noStyle />, { wrapper: LinkTestWrapper });

  expect(asFragment()).toMatchSnapshot();
});
