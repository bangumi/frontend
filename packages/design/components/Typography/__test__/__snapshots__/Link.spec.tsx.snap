// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render external link 1`] = `
<DocumentFragment>
  <a
    class="bgm-link"
    href="https://test.com"
  />
</DocumentFragment>
`;

exports[`should render internal link 1`] = `
<DocumentFragment>
  <a
    class="bgm-link"
    href="/a"
  />
</DocumentFragment>
`;

exports[`should render no-style link 1`] = `
<DocumentFragment>
  <a
    href="/a"
  />
</DocumentFragment>
`;
